src/vsc_r_api.ts

```diff
@@ -0,0 +1,374 @@
+    const newRange = parseRange(x, targetDocument);
+    const newSelection = new Selection(newRange.start, newRange.end);
+    return (newSelection);
+  });
+
+  editor.selections = selectionObjects;
+}
+
+export async function documentSave(id: string) {
+  const target = findTargetUri(id);
+  const targetDocument = await workspace.openTextDocument(target);
+  await targetDocument.save();
+}
+
+export async function documentSaveAll() {
+  await workspace.saveAll();
+}
+
+export function projectPath() {
+
+  if (typeof workspace.workspaceFolders !== 'undefined') {
+    // Is there a root folder open?
+
+    if (workspace.workspaceFolders.length === 1) {
+      // In single root common case, this will always work.
+      return {
+        path: workspace.workspaceFolders[0].uri.path
+      };
+    } else if (workspace.workspaceFolders.length > 1) {
+      // In less common multi-root folder case is a bit tricky. If the active
+      // text editor has scheme 'untitled:' (is unsaved), then
+      // workspace.getWorkspaceFolder() won't be able to find its Uri in any
+      // folder and will return undefined.
+      const currentDocument = getLastActiveTextEditor().document;
+      const currentDocFolder = workspace.getWorkspaceFolder(currentDocument.uri);
+      if (typeof currentDocFolder !== 'undefined') {
+        return {
+          path: currentDocFolder.uri.path
+        };
+      }
+    }
+  }
+
+  // if we got to here either: 
+  //   - the workspaceFolders array was undefined (no folder open)
+  //   - the activeText editor was an unsaved document, which has undefined workspace folder.
+  // return undefined and handle with a message in R.
+  return {
+    path: undefined
+  };
+}
+
+export async function documentNew(text: string, type: string, position: number[]) {
+  const documentUri = Uri.parse('untitled:' + path.join(projectPath().path, 'new_document.' + type));
+  const targetDocument = await workspace.openTextDocument(documentUri);
+  const edit = new WorkspaceEdit();
+  const docLines = targetDocument.lineCount;
+  edit.replace(documentUri,
+    targetDocument.validateRange(new Range(
+      new Position(0, 0),
+      new Position(docLines + 1, 0)
+    )),
+    text);
+
+  workspace.applyEdit(edit).then(async () => {
+    const editor = await window.showTextDocument(targetDocument);
+    editor.selections = [new Selection(
+      parsePosition(position, targetDocument),
+      parsePosition(position, targetDocument)
+    )];
+  });
+}
+
+// interface
+// represents addins in a QuickPick menu
+interface AddinItem extends QuickPickItem {
+  binding: string;
+  package: string;
+}
+
+let addinQuickPicks: AddinItem[] = undefined;
+
+export async function getAddinPickerItems() {
+
+  if (typeof addinQuickPicks === 'undefined') {
+    const addins: any[] = await fs.readJSON(path.join(sessionDir, 'addins.json'));
+    const addinItems = addins.map((x) => {
+      return {
+        alwaysShow: true,
+        description: x.description,
```

Comment:
It might look better if we use the following:

```
description: `{${x.package}}`,
detail: x.description
```

and the addin picker will look like 

<img width="603" alt="image" src="https://user-images.githubusercontent.com/4662568/95659112-6f26c380-0b51-11eb-8ad9-bf0e1674b8ec.png">

Reply:
Yup I agree.